### PR TITLE
Added notifications for bytes read and written during redis requests.

### PR DESCRIPTION
**Pull request by simongui**
These changes collect the bytes for outgoing redis requests and incoming responses and publish notifications with the byte size collected.

### CODE REVIEW

File: lib/redis/client.rb [hunk 1]
```diff
@@ -247,14 +247,30 @@ def read
...
         value
       end
     end
 
     def write(command)
-      io do
-        @pending_reads += 1
-        connection.write(command)
+      bytes = 0
+      for cmd in command
```
Comment (fw42) [inline lib/redis/client.rb:265]: `for` loops are bad style in Ruby.. try to prefer enumerators and `#each`, i.e.

``` ruby
command.each do |cmd|
  # ...
end
```

**Comment by fw42** [comment]
Hmm.. why not hook into `build_command` (lib/redis/connection/command_helper.rb) or into `Connection#write`?

It's a bit far fetched but right now someone could change the way commands to the client are translated into bytes on the socket (and use more or fewer bytes) but your code would still count the same number of bytes. I see this as being on the wrong level of abstraction (you add "number of bytes written to the socket" on a layer that knows nothing about sockets).

File: lib/redis/client.rb [hunk 2]
```diff
@@ -247,14 +247,30 @@ def read
       io do
         value = connection.read
         @pending_reads -= 1
+
+        bytes = 0
+        if !value.nil?
+          bytes += value.size
+        end
+
+        ActiveSupport::Notifications.instrument('redis-rb.read.bytes', extra: bytes) do
```
Comment (fw42) [inline lib/redis/client.rb:256]: ActiveSupport is not a dependency of this gem. If anyone uses this gem in an application that doesn't depend on ActiveSupport already, this code will break.

**Comment by fw42** [comment]
I think the way I would implement this is by implementing a new Redis Connection driver (a subclass of Hiredis or whatever we use right now) and make that new driver count the data that goes in / comes out of the socket (maybe by creating a new "CountingSocket" class that gets injected into the driver).

**Comment by simongui** [comment]
@fw42 The Redis protocol is plain text and meant to be usable over Telnet etc. As far as I can tell (correct me if I'm wrong, my Ruby isn't strong yet) the hiredis C driver accepts strings and there's no socket or byte visibility at the Ruby layer because that's all done in C. If we want bytes at the socket level, changes in C are going to be required. 

Do you think we should go through with that?

**Comment by fw42** [comment]
> Do you think we should go through with that?

hm no probably not

**Comment by fw42** [comment]
ok then go with your approach... although I feel like this should be a monkey-patch in shopify core, not a modification to the gem.. that would also get around the ActiveSupport dependency thing I mentioned...

**Comment by simongui** [comment]
@fw42 What do you mean by monkey-patch? Shopify core you mean the https://github.com/shopify/shopify repo?

**Comment by simongui** [comment]
FYI, the Shopify side of my changes are similar to this to stay consistent with some of the other ways we collect metrics. That context may have been missing since it's not in this pull request.

https://github.com/Shopify/shopify/blob/9511083e6d315abfbd005413f8287525603881ec/lib/memcache_counter.rb

**Comment by simongui** [comment]
Closing because we decided on a different approach.

**Comment by fw42** [comment]
can we delete this repository now?

**Comment by simongui** [comment]
Right now we are deploying from a @byroot branch I believe. This repo should be up to date now. Might be a good idea to start deploying from here at some point.

**Comment by fw42** [comment]
if it's up to date with upstream then we should just deploy from upstream and get rid of our fork

**Comment by byroot** [comment]
My PR haven't been merged upstream yet: https://github.com/redis/redis-rb/pull/509

**Comment by simongui** [comment]
Might be nicer to deploy from shopify/redis-rb though.

**Pull request closed without merging**